bigflow/workflow.py

```diff
@@ -29,8 +30,36 @@ def daily_start_time(start_time: dt.datetime) -> dt.datetime:
     return start_time.replace(hour=0, minute=0, second=0, microsecond=0) - td
 
 
+class JobContext(typing.NamedTuple):
+    runtime_raw: typing.Optional[str]
+    runtime: typing.Optional[dt.datetime]
+    env: typing.Optional[str]
+    workflow: typing.Optional['Workflow']
+    workflow_id: typing.Optional[str]
+    # TODO: add unique 'workflow execution id' (for tracing/logging)
+    # TODO: add some 'workflow-level' configuration here
+
+
+class Job(abc.ABC):
+    """Base abstract class for bigflow.Jobs.  It is recommended to inherit all your jobs from this class."""
+
+    @property
+    @abc.abstractmethod
+    def id(self) -> str:
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def execute(self, context: JobContext):
```

Comment:
I'm not sure if we want to change the interface so drastically. 

* Most of the jobs won't use all of these parameters, only the runtime parameter.
* We have many jobs to change in our analytics, even though they don't need the additional context.
* It sets the bar a bit higher at the beginning (single parameter `runtime` is easier to understand than the rich `JobContext`).

I suggest something in the middle:
```
@abc.abstractmethod
def run(self, runtime, additional_context: Optional[JobContext]):
    pass
```

In that case, if a user doesn't need additional context, a job can be very simple:
```
class MyJob:
    id = 'my_job'
    
    def run(self, runtime):
        pass
```

And if a user needs some additional context info, then takes an additional argument:
```
class MyJob:
    id = 'my_job'
    
    def run(self, runtime, additional_context):
        pass
```

Reply:
Hm, I thought about this option. 
However it requires slightly more runtime magic - inspecting method signature (checking number of args).
Python doesn't have overloading by number of arguments, so it looks like attempt to use idiom from other languages.
IMO It makes harder to understand the code for external contributors and users.

Reply:
Also the idea was to add parsed version of 'runtime', so user don't need to deal with parsing, different format (and care about this format at all).  That is why 'context' include 'runtime' and 'runtime_raw'.

Reply:
Well, that's true that doing Java from Python is not the best idea. The second argument is also true (though these operations are super simple, and we need to maintain, test, and document them).

I suggest that in the first version we keep the execute method out of the public interface, for the internal utils like dataproc job etc. **Making it public and recommended requires a lot of work in the documentation and our projects**. What is more, we don't need it in our codebase at this moment.

@bartoszwalacik @iwko what do you think?